src/lua/uri.lua

```diff
@@ -62,12 +68,11 @@ local uri_set_stash = buffer.ffi_stash_new('struct uri_set')
 local uri_set_stash_take = uri_set_stash.take
 local uri_set_stash_put = uri_set_stash.put
 
-local function escape_pattern(text)
-    return text:gsub("([^%w])", "%%%1")
-end
-
 local unreserved = function(str)
-    return "^[" .. escape_pattern(str) .. "]$"
+    assert(type(str) == "string")
+    local c_unreserved = ffi.new("char[?]", #str + 1)
+    ffi.copy(c_unreserved, str, #str)
+    return c_unreserved
 end
```

Comment:
The simple Lua implementation makes all needed steps to prepare `unreserved` within this function. It is logical.

The Lua/FFI implementation doesn't do that and it looks counter-intuitive. We create some intermediate representation here to post-process it in `escape`/`unescape` (`build_unreserved_symbols_table`). It requires extra actions and an extra global variable.

I think we should build the necessary table right here and store it in the returned object.